New Examples for Decentralized Optimization

sebcolla opened the pull request:
Hello Adrien,

I have created some new examples that I use in a journal paper that I will submit soon (probably in early April)
I have also updated some parts of the file "matrix.m".
I have run all examples via 'test_install'.

Best,

Sébastien

sebcolla committed:
Examples PEP for DGD

Codes used for obtaining the results of the following paper:
Colla, Sebastien, and Julien M. Hendrickx. "Automated Worst-Case Performance Analysis of Decentralized Gradient Descent." (2021)

sebcolla committed:
changes for pull request

sebcolla committed:
add source of the paper for Decentralized Examples

sebcolla committed:
PEP for Decentralized Optimization

New feature for PESTO: possibility of adding new type of constraints (LMIs)
New examples: Decentralized Gradient Descent (DGD)

sebcolla committed:
Update pep output (dualvalues and dualnames)

sebcolla committed:
Update Examples Decentralized Optimization

sebcolla committed:
Update Decentralized Optimization Examples

sebcolla committed:
Merge remote-tracking branch 'upstream/master'

sebcolla committed:
update pep from PESTO git

sebcolla committed:
Merge branch 'AdrienTaylor:master' into master

sebcolla committed:
update PESTO for decentralized optimization

sebcolla committed:
Update DGD example

Update DGD example with the new "user friendly" interface for decentralized optimization

sebcolla committed:
Update UserGuide about new features for decentralized optimization

sebcolla committed:
Update A_DecentralizedGradientDescent

sebcolla committed:
Update A_DecentralizedGradientDescent

sebcolla committed:
subfolder with auxiliary functions

I have also added a line in the install of pesto to save the added path

sebcolla committed:
update A_DecentralizedGradientDescent for PR

sebcolla committed:
update userguide authors

sebcolla committed:
minor changes for PR

sebcolla committed:
update toolbox from github

sebcolla committed:
update from github

sebcolla committed:
Merge branch 'master' of https://github.com/AdrienTaylor/Performance-Estimation-Toolbox into AdrienTaylor-master

sebcolla committed:
merge PointSaga example

sebcolla committed:
Merge branch 'AdrienTaylor-master'
	erge to have last PESTO version and examples

sebcolla committed:
remove unnecessary files

sebcolla committed:
merge

sebcolla committed:
merge

sebcolla committed:
update examples Decentralized Optimization

sebcolla committed:
update and add examples in decentralized optimization

sebcolla committed:
updating and adding examples in decentralized optimization

sebcolla committed:
update examples for decentralized optimization

sebcolla committed:
update examples for decentralized optimization

sebcolla committed:
samll updates for DIGing example

update of the theoretical worst-case bound

AdrienTaylor commented on Examples/11_Decentralized Optimization Methods/B_DIGing.m:
```diff
@@ -0,0 +1,142 @@
+returnOpt = 0;
+[Fi,Fav,~,~] = P.DeclareMultiFunctions(fctClass,fctParam,N,returnOpt);
+[xs,Fs] = Fav.OptimalPoint(); 
+P.AddConstraint(xs^2 == 0); % we can set xs = 0, without loss of generality
```
This is numerically quite not good (it kills the existence of a Slater point), and can really deteriorate the performance of numerical SDP solvers -> please avoid it (by using xs at appropriate places)

AdrienTaylor commented on Examples/11_Decentralized Optimization Methods/B_DIGing.m:
```diff
@@ -0,0 +1,142 @@
+[Fi,Fav,~,~] = P.DeclareMultiFunctions(fctClass,fctParam,N,returnOpt);
+[xs,Fs] = Fav.OptimalPoint(); 
+P.AddConstraint(xs^2 == 0); % we can set xs = 0, without loss of generality
+P.AddConstraint(Fs == 0);   % we can set Fs = 0, without loss of generality
```
It is not as critical as the previous point but still creates a variable which is constrained to be zero -> can we avoid that?

AdrienTaylor commented on Examples/11_Decentralized Optimization Methods/C_DIGing_rate.m:
```diff
@@ -0,0 +1,146 @@
+returnOpt = 0;
+[Fi,Fav,~,~] = P.DeclareMultiFunctions(fctClass,fctParam,N,returnOpt);
+[xs,Fs] = Fav.OptimalPoint(); 
+P.AddConstraint(xs^2 == 0); % we can set xs = 0, without loss of generality
```
same comment :)

AdrienTaylor commented on Examples/11_Decentralized Optimization Methods/C_DIGing_rate.m:
```diff
@@ -0,0 +1,146 @@
+[Fi,Fav,~,~] = P.DeclareMultiFunctions(fctClass,fctParam,N,returnOpt);
+[xs,Fs] = Fav.OptimalPoint(); 
+P.AddConstraint(xs^2 == 0); % we can set xs = 0, without loss of generality
+P.AddConstraint(Fs == 0);   % we can set Fs = 0, without loss of generality
```
same comment :

AdrienTaylor commented on Examples/11_Decentralized Optimization Methods/D_AccDNGD.m:
```diff
@@ -0,0 +1,147 @@
+returnOpt = 0;
+[Fi,Fav,~,~] = P.DeclareMultiFunctions(fctClass,fctParam,N,returnOpt);
+[xs,Fs] = Fav.OptimalPoint(); 
+P.AddConstraint(Fs == 0);   % we can set Fs = 0, without loss of generality
```
same comment :)

AdrienTaylor commented on Examples/11_Decentralized Optimization Methods/D_AccDNGD.m:
```diff
@@ -0,0 +1,147 @@
+X(:,1) = P.MultiStartingPoints(N,equalStart);
+V(:,1) = X(:,1); 
+Y(:,1) = X(:,1);
+P.AddConstraint(X{1,1}^2 == 0);         % starting with x0 = v0 = y0 = 0, for all the agents
```
same comment :) (please check at other places as well)

AdrienTaylor reviewed (changes requested):
Petits changements demandés. (listés plus haut)

- les tests sont OK
- les outputs ne sont pas toutes très propres pour les cas décentralisés; p-e les nettoyer un peu//faire en sorte qu'ils passent sur une ligne? (c'est une demande mal définie, mais en tout cas tu peux faire des messages plus courts sur chaque ligne)

Merci!

sebcolla committed:
modifications for pull request

update the outputs to make them shorter.
Remove equality constraints to avoid rank deficiency and improve numerical SDP resolution.

AdrienTaylor reviewed (approved):
Hello Sébastien,

I've tested & checked your code; I think it is OK.
